ThirdPartyModules/Route/Route.php

```diff
@@ -41,76 +78,184 @@ class Route
      *
      * @return array either changed or unchanged cart array
      */
-    public function filterCartBeforeLegacyShippingAndTax($cart)
+    public function filterCartBeforeSplitShippingAndTax($cart)
     {
+        $parentQuoteId = $cart['order_reference'];
         $cart['items'] = array_filter(
             $cart['items'],
-            function ($item) {
-                return $item['reference'] !== RouteDataHelper::ROUTE_FEE;
+            function ($item) use($parentQuoteId) {
+                $this->saveRouteFeeEnabled($item['reference'], $parentQuoteId, '1');
+                return $item['reference'] !== self::ROUTE_PRODUCT_ID;
             }
         );
+        
         return $cart;
     }
+    
+    /**
+     * Filter the transaction before Create Order functionality is executed
+     * Remove a dummy product representing the Route Fee additional total, also update cache if Route fee is enabled
+     *
+     * @see \Bolt\Boltpay\Model\Api\CreateOrder::createOrder
+     *
+     * @param array $cart the transaction object from Bolt
+     *
+     * @return array either changed or unchanged $transaction object
+     */
+    public function filterCartBeforeCreateOrder($transaction)
+    {
+        $parentQuoteId = $transaction->order->cart->order_reference;
+        $transaction->order->cart->items = array_filter(
+            $transaction->order->cart->items,
+            function ($item) use($parentQuoteId) {
+                $this->saveRouteFeeEnabled($item->reference, $parentQuoteId, '1');
+                return $item->reference !== self::ROUTE_PRODUCT_ID;
+            }
+        );
+        
+        return $transaction;
+    }
+    
+    /**
+     * Filter the item before adding to cart.
+     *
+     * @see \Bolt\Boltpay\Model\Api\UpdateCart::execute
+     * 
+     * @param bool $flag
+     * @param array $addItem
+     * @param \Magento\Checkout\Model\Session $checkoutSession
+     *
+     * @return bool if return true, then move to next item, if return false, just process the normal execution.
+     */
+    public function filterAddItemBeforeUpdateCart($result, $addItem, $checkoutSession)
+    {
+        if ($addItem['product_id'] == self::ROUTE_PRODUCT_ID) {
+            $checkoutSession->setInsured(true);
+            $this->saveRouteFeeEnabled(self::ROUTE_PRODUCT_ID, $checkoutSession->getQuote()->getBoltParentQuoteId(), '1');
+            return true;
+        }
+        
+        return $result;
+    }
+    
+    /**
+     * Filter the item before removing from cart.
+     *
+     * @see \Bolt\Boltpay\Model\Api\UpdateCart::execute
+     * 
+     * @param bool $flag
+     * @param array $removeItem
+     * @param \Magento\Checkout\Model\Session $checkoutSession
+     *
+     * @return bool if return true, then move to next item, if return false, just process the normal execution.
+     */
+    public function filterRemoveItemBeforeUpdateCart($result, $removeItem, $checkoutSession)
```

Comment:
`filterRemoveItemBeforeUpdateCart` and `filterAddItemBeforeUpdateCart` can be joined into one method with additional `setInsured` flag parameter. Anyway, it's a matter of preference, not a real requirement.

Reply:
Fixed